File: back/src/app/api/fetch_salaries.rs
use std::{sync::Arc, collections::HashMap};

use crate::{
    domain::{models::{Salary, salary::Order}, use_cases},
    infra::SalaryRepository,
};
use axum::{extract::{State, Query}, http::StatusCode, Json};
use chrono::NaiveDate;
use serde::Serialize;

#[derive(Serialize)]
pub struct Response {
    company: String,
    title: Option<String>,
    location: String,
    compensation: i32,
    date: NaiveDate,
    level: Option<String>,
    company_xp: Option<i32>,
    total_xp: Option<i32>,
}

impl From<Salary> for Response {
    fn from(salary: Salary) -> Self {
        Self {
            company: salary.company.into(),
            title: salary.title.map(|title| title.into()),
            location: salary.location.into(),
            compensation: salary.compensation.into(),
            date: salary.date.into(),
            level: salary.level.map(|level| level.into()),
            company_xp: salary.company_xp.map(|company_xp| company_xp.into()),
            total_xp: salary.total_xp.map(|total_xp| total_xp.into()),
        }
    }
}

type Error = (StatusCode, &'static str);

pub async fn fetch_salaries(
    State(repo): State<Arc<dyn SalaryRepository>>,
    Query(params): Query<HashMap<String, String>>
) -> Result<Json<Vec<Response>>, Error> {
    let order = Order::from(params);

    match use_cases::fetch_salaries(repo, order).await {
        Ok(salaries) => Ok(salaries
            .into_iter()
            .map(|salary| salary.into())
            .collect::<Vec<Response>>()
            .into()),
        Err(use_cases::fetch_salaries::Error::Unknown(str)) => {
            Err((StatusCode::INTERNAL_SERVER_ERROR, str))
        }
    }
}

File: src/app/api/fetchSalaries.ts
import { Context } from "hono";
import * as lib from "@domain";
import { SalaryRepository } from "@infra";
import { Maybe, Result } from "@utils";

export async function fetchSalaries(
    c: Context,
    salaryRepo: SalaryRepository
) {
    return Result.match(await lib.fetchSalaries({
        key: lib.Key.default(),
        direction: lib.Direction.default()
    }, salaryRepo), {
        onOk: (salaries) => c.json(salaries.map(salary => ({
            company: lib.Company.toString(salary.company),
            title: Maybe.toNullable(Maybe.map(salary.title, lib.Title.toString)),
            location: lib.Location.toString(salary.location),
            compensation: lib.Compensation.toNumber(salary.compensation),
            date: lib.SalaryDate.toDate(salary.date),
            level: Maybe.toNullable(Maybe.map(salary.level, lib.Level.toString)),
            company_xp: Maybe.toNullable(Maybe.map(salary.companyXp, lib.Xp.toNumber)),
            total_xp: Maybe.toNullable(Maybe.map(salary.totalXp, lib.Xp.toNumber)),
        }))),
        onErr: () => c.json([], 500)
    });
}
